packages/ruleset/src/functions/use-date-based-format.js

```diff
@@ -0,0 +1,408 @@
+      );
+
+      return value;
+    }
+  }
+
+  // Look for examples at different paths up the logical path.
+  const examplePath = [...logicalPath];
+
+  // Look at each level of hierarchy of the example path, INCLUDING examples on
+  // primary schemas, which will have an example path of an empty array. For
+  // that reason, we can't loop through the example path elements, as we would
+  // miss the iteration with no elements in the array. Using a simple for-loop
+  // that executes one more time than there are segments in the example path,
+  // we execute this logic for the empty path scenario as well, and thereby
+  // properly handle primary schemas. The length of the example path is used to
+  // initialize the counter, rather than in the conditional, since the array
+  // will change length as we progress.
+  for (let i = examplePath.length; i > -1; i--) {
+    logger.debug(
+      `${ruleId}: Looking at example at logical path ${logicalPathForLogger(
+        examplePath
+      )}`
+    );
+    const value = findValueInExamples(
+      examples[examplePath.join('.')],
+      logicalPath,
+      examplePath
+    );
+
+    // If we find a value, go ahead and return.
+    if (value !== undefined) {
+      logger.debug(
+        `${ruleId}: example value found in schema example at logical path: ${logicalPathForLogger(
+          examplePath
+        )}: ${value}`
+      );
+
+      return value;
+    }
+
+    // Pop the last element to keep looking up the path.
+    examplePath.pop();
+  }
+
+  // This will return `undefined` if we never find a value;
+  logger.debug(
+    `${ruleId}: no example value found for schema at logical path: ${logicalPathForLogger(
+      logicalPath
+    )}`
+  );
+}
+
+function findValueInExamples(examples, logicalPath, examplePath) {
+  if (examples && examples.length) {
+    // First thing is to find the relevant segment of the logical path relative
+    // to the example path, which should be the first element where they differ.
+    const index = logicalPath.findIndex((prop, i) => prop !== examplePath[i]);
+
+    // This means that the logical paths are the same, i.e. the examples are
+    // primitives defined directly on the schema.
+    if (index === -1) {
+      // Return the first defined example.
+      return examples.find(e => e !== undefined);
+    }
+
+    const relativeLogicalPath = logicalPath.slice(index);
+    for (const example of examples) {
+      const value = getObjectValueAtPath(example, relativeLogicalPath);
+
+      // If we find a value, go ahead and return it.
+      if (value !== undefined) {
+        return value;
+      }
+    }
+  }
+
+  return undefined;
+}
+
+// This function takes an object, as well as a path to a specific value, and
+// recursively parses the object looking for the value at that path. If it
+// finds one, the value will be returned. If not, the function will return
+// `undefined`.
+function getObjectValueAtPath(obj, pathToValue) {
```

Comment:
Recursion is probably overkill here — a loop should be fine.

Reply:
I'm surprised you didn't suggest `reduce` here 🙂 

I felt the recursion was cleaner than when I had the implementation as a loop, but I'll consider adjusting it.

Reply:
`reduce` ended up working best, fixed